src/JavascriptVM/JavascriptVM.tsx

```diff
@@ -310,9 +314,21 @@ export const VMProvider: FunctionComponent = ({ children }) => {
             challengeListener.current.onStop();
           }
 
+          // Before we start the challenge listener we mark the challenge
```

Comment:
One disadvantage of this approach is that it we might "loose" the state of a session when it is currently marked as failure.  E.g. say a lesson has been marked as a failure, and the users starts over and runs the code again, and this time the code doesn't produce anything that triggers a 'failure' event (e.g. the robot did not drive into a bad zone) then the lesson state goes from 'failure' to 'pending'.  Maybe that's okay.

Alternatively, instead of resetting the state here, we could pass the current state  to the ChallengeActionsImpl object, and have the ChallengeActionsImpl figure out the details of how to correctly update the slice.  The ChallengeActionsImpl gets notified of all the events happening during the current run, together with the status of the challenge at the beginning of the run it can then update the slice.